File: sdk/python/trust_router_client.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import requests


@dataclass(frozen=True)
class TrustRouterClient:
    """Small sidecar client for wrapping tool-calling loops."""

    base_url: str = "http://127.0.0.1:7878"
    tenant: str = "yc-demo"
    timeout_seconds: float = 5.0
    api_key: str = "trust-router-demo-key"

    def route(self, start: str, goal: str) -> dict[str, Any]:
        response = requests.get(
            f"{self.base_url.rstrip('/')}/route",
            params={"tenant": self.tenant, "start": start, "goal": goal},
            timeout=self.timeout_seconds,
            headers={"X-API-Key": self.api_key},
        )
        response.raise_for_status()
        return response.json()

    def report_result(self, node: str, success: bool, latency_ms: int) -> dict[str, Any]:
        response = requests.post(
            f"{self.base_url.rstrip('/')}/result",
            params={
                "tenant": self.tenant,
                "node": node,
                "success": str(success).lower(),
                "latency_ms": latency_ms,
            },
            timeout=self.timeout_seconds,
            headers={"X-API-Key": self.api_key},
        )
        response.raise_for_status()
        return response.json()


def langchain_style_example() -> None:
    router = TrustRouterClient()

    # In a LangChain-style loop, ask Trust Router for the deterministic path
    # before letting the agent spend tokens deciding routine tool order.
    decision = router.route(start="start", goal="done")
    if decision["decision"] == "escalate":
        # Fall back to the LLM only when the graph has no feasible route.
        return

    for tool_name in decision["path"][1:]:
        try:
            # Replace this comment with: tool_registry[tool_name].invoke(...)
            latency_ms = 0
            router.report_result(tool_name, success=True, latency_ms=latency_ms)
        except Exception:
            router.report_result(tool_name, success=False, latency_ms=30_000)
            raise

File: sdk/typescript/trustRouterClient.ts
export type RouteDecision =
  | {
      decision: "routed";
      path: string[];
      total_cost: number;
      cache_hit: boolean;
    }
  | {
      decision: "escalate";
      reason: string;
      failed_or_blocked_nodes: string[];
    };

export type PlanDecision =
  | {
      decision: "execute";
      schema_version: number;
      path: string[];
      total_cost: number;
      cache_hit: boolean;
      steps: string[];
    }
  | {
      decision: "recover";
      reason: string;
      failed_or_blocked_nodes: string[];
      recovery_steps: string[];
    };

export type ResultResponse = {
  ok: boolean;
};

export type TrustRouterClientOptions = {
  baseUrl?: string;
  tenant?: string;
  timeoutMs?: number;
  apiKey?: string;
};

export class TrustRouterClient {
  private readonly baseUrl: string;
  private readonly tenant: string;
  private readonly timeoutMs: number;
  private readonly apiKey: string;

  constructor(options: TrustRouterClientOptions = {}) {
    this.baseUrl = (options.baseUrl ?? "http://127.0.0.1:7878").replace(/\/+$/, "");
    this.tenant = options.tenant ?? "yc-demo";
    this.timeoutMs = options.timeoutMs ?? 5_000;
    this.apiKey = options.apiKey ?? "trust-router-demo-key";
  }

  async route(start: string, goal: string): Promise<RouteDecision> {
    const params = new URLSearchParams({
      tenant: this.tenant,
      start,
      goal,
    });

    return this.request<RouteDecision>(`/route?${params.toString()}`, {
      method: "GET",
    });
  }

  async plan(start: string, goal: string): Promise<PlanDecision> {
    const params = new URLSearchParams({
      tenant: this.tenant,
      start,
      goal,
    });

    return this.request<PlanDecision>(`/plan?${params.toString()}`, {
      method: "GET",
    });
  }

  async reportResult(
    node: string,
    success: boolean,
    latencyMs: number,
  ): Promise<ResultResponse> {
    const params = new URLSearchParams({
      tenant: this.tenant,
      node,
      success: String(success),
      latency_ms: String(latencyMs),
    });

    return this.request<ResultResponse>(`/result?${params.toString()}`, {
      method: "POST",
    });
  }

  private async request<T>(path: string, init: RequestInit): Promise<T> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        ...init,
        headers: {
          ...(init.headers ?? {}),
          "X-API-Key": this.apiKey,
        },
        signal: controller.signal,
      });
      const body = await response.text();

      if (!response.ok) {
        throw new Error(`Trust Router request failed: ${response.status} ${body}`);
      }

      return JSON.parse(body) as T;
    } finally {
      clearTimeout(timeout);
    }
  }
}
